lib/configSchema.js

```diff
@@ -0,0 +1,92 @@
+'use strict';
+
+const functionNamePattern = '^[a-zA-Z0-9-_]+$';
+
+const schema = {
+  type: 'object',
+  properties: {
+    service: {
+      type: 'object',
+      properties: {
+        name: { pattern: '^[a-zA-Z][0-9a-zA-Z-]+$' },
+        awsKmsKeyArn: { pattern: '^arn:(aws[a-zA-Z-]*)?:kms:[a-z0-9-]+-\\d+:\\d{12}:[^\\s]+$' },
+      },
+      additionalProperties: false,
+    },
+
+    custom: {
+      type: 'object',
+      properties: {},
+      required: [],
+      additionalProperties: true, // user is free to add properties that he or she wants
+    },
+    plugins: { type: 'array', items: { type: 'string' } },
+
+    // TODO: Describe resources schema
+    resources: { type: 'object' },
+
+    functions: {
+      type: 'object',
+      patternProperties: {
+        [functionNamePattern]: {
+          type: 'object',
+          properties: {
+            name: { type: 'string' }, // name property is added by service class
+            events: {
+              type: 'array',
+              items: {}, // event plugins push their schemas to this array
```

Comment:
I just realized that we have one subtle leak here, e.g. if custom provider defines its schema, but wouldn't define any events, then it'll come unnoticed, and any eventual events as configured in service config would be processed without any validation and notice.

It appears there's no clean way to address it with JSON schema means, but I think we can achieve the desired behavior by hacking schema a bit.

I propose to assign below schema to `items` here:

```javascript
	anyOf: [
		{
			type: 'object',
			properties: { __schemaWorkaround__: { const: null } },
			required: ['__schemaWorkaround__'],
			additionalProperties: false,
		},
	],
```

And imo it's fine to leave this item even if some events are configured for a function.

Above will ensure that unrecognized events are reported if provider extended the schema without event schemas.

Then in `relaxSchema` I believe we will need to revert this hack, by deleting  `items.anyOf`  (still, as we also use it now for temporary AWS relaxation, we should delete it only if it's length is `1`)

Reply:
Let me share few thoughts on that. If we implement this then the provider author will be forced to implement all the schemas for each event, at least relaxed ones. Yes, it's authors responsibility, but why push him or her to provide all shemas at once? The author may gradually impement schemas, first for plugin. 

The proposed solutions would also add complexity to existing code.

Please confirm that you still want this

Reply:
On the other hand it would be nice to have a reminder that would enourage provider author to implement schemas for events too

Reply:
Yes, I think it's fine to do that. Note that provider implementer may take same approach as we did here with AWS, by implementing at first permissive schemas for each supported events, and then in future completing them step by step.

Reply:
It's also the only way to pick typos in event names, which I think it's quite important to signal

Reply:
Yes, it is fine to do that. But is it necessary? We are going to ship a big feature, so something may come up that we would not expect. For this reason, I think it's better to have a more simple codebase. And the proposed solution adds more complexity rather than value. 

Should we maybe include other participants who would make tests of our feature? It's a big feature and I believe it's good to have the support of other members.

Reply:
Sure, but do you see any risk with this addition?

To me it feels that it addresses known issue and there's no side effect, also implementation wise it's purely about injecting proposed schema in.

Reply:
I don't see any risk. 

Thanks for the explanation, I'll implement this too